prov/tcp/src/xnet.h

```diff
@@ -467,12 +467,67 @@ struct xnet_xfer_entry {
 	char			msg_data[];
 };
 
+struct xnet_mplex_av {
+	struct util_av		util_av;
+	struct dlist_entry	subav_list;
+	struct ofi_genlock	lock;
+};
+
 struct xnet_domain {
 	struct util_domain		util_domain;
 	struct xnet_progress		progress;
 	enum fi_ep_type			ep_type;
+
+	// When an application requests FI_THREAD_COMPLETION
+	// the assumption is that the domain will be used
+	// across multiple threads.
+	//
+	// The xnet progress engine is optimized for single
+	// threaded performance, so instead of reworking the
+	// progress engine, likely losing single threaded
+	// performance, multiplex the domain into multiple
+	// subdomains for each ep. This way, each ep, with
+	// the assumption that the application wants to
+	// progress an ep per thread, can have it's own
+	// progress engine and avoid having a single
+	// synchronization point among all eps.
+	struct fi_info		*subdomain_info;
+	struct ofi_genlock	subdomain_list_lock;
+	struct dlist_entry	subdomain_list;
 };
 
+int xnet_multiplex_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
+			   struct fid_av **fid_av, void *context);
+
+int xnet_domain_open(struct fid_fabric *fabric, struct fi_info *info,
+		     struct fid_domain **domain, void *context);
+
+static inline
+int xnet_create_subdomain(struct xnet_domain *domain, struct xnet_domain **subdomain)
+{
+	int ret;
+	struct fid_domain *subdomain_fid;
+
+	ret = xnet_domain_open(&domain->util_domain.fabric->fabric_fid,
+			       domain->subdomain_info,
+			       &subdomain_fid, NULL);
+	if (ret)
+		return ret;
+
+	*subdomain = container_of(subdomain_fid, struct xnet_domain,
+				  util_domain.domain_fid);
+
+	return FI_SUCCESS;
+}
+
+int xnet_domain_multiplexed(struct fid_domain *domain_fid);
+
+int xnet_multiplex_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
+			   struct fid_av **fid_av, void *context);
+
+int xnet_domain_open(struct fid_fabric *fabric, struct fi_info *info,
+		     struct fid_domain **domain, void *context);
+
```

Comment:
xnet_domain_open is the original default, we add the check to see if the domain is multiplexed there.  it already existed, just moved a few lines (see line 547 on the old version of this file)

Reply:
Oh I'm talking about lines 499-503 in the new file - it looks like you added it twice up here. But I don't think they need to move because I think they had to move to call in the new static inline for create_subdomains which never called. So think this whole block can be dropped except the xnet_domain_multiplexed call. At least that's how I'm reading it